nimbus/src/enrollment.rs

```diff
@@ -311,55 +758,820 @@ mod tests {
+        assert_eq!(events[0].experiment_slug, exp.slug);
+        assert_eq!(events[0].branch_slug, "control");
+        assert_eq!(events[0].reason, Some("optout".to_owned()));
+        assert_eq!(
+            events[0].change,
+            EnrollmentChangeEventType::Disqualification
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_evolver_experiment_update_enrolled_then_experiment_paused() -> Result<()> {
+        let mut exp = get_test_experiments()[0].clone();
+        exp.is_enrollment_paused = true;
+        let (nimbus_id, app_ctx, aru) = local_ctx();
+        let evolver = enrollment_evolver(&nimbus_id, &app_ctx, &aru);
+        let mut events = vec![];
+        let enrollment_id = Uuid::new_v4();
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug.clone(),
+            status: EnrollmentStatus::Enrolled {
+                enrollment_id,
+                branch: "control".to_owned(),
+                reason: EnrolledReason::Qualified,
+            },
+        };
+        let enrollment = evolver
+            .evolve_enrollment(
+                true,
+                Some(&exp),
+                Some(&exp),
+                Some(&existing_enrollment),
+                &mut events,
+            )?
+            .unwrap();
+        dbg!(&enrollment.status);
+        if let EnrollmentStatus::Enrolled {
+            reason: EnrolledReason::Qualified,
+            enrollment_id: new_enrollment_id,
+            ..
+        } = enrollment.status
+        {
+            assert_eq!(new_enrollment_id, enrollment_id);
+        } else {
+            panic!("Wrong variant!");
         }
+        assert!(events.is_empty());
         Ok(())
     }
 
-    fn get_experiment_enrollments(db: &Database) -> Result<Vec<ExperimentEnrollment>> {
-        db.collect_all::<ExperimentEnrollment>(StoreId::Enrollments)
+    #[test]
+    fn test_evolver_experiment_update_enrolled_then_bucketing_changed() -> Result<()> {
+        let mut exp = get_test_experiments()[0].clone();
+        exp.bucket_config.count = 0; // Make the experiment bucketing fail.
+        let (nimbus_id, app_ctx, aru) = local_ctx();
+        let evolver = enrollment_evolver(&nimbus_id, &app_ctx, &aru);
+        let mut events = vec![];
+        let enrollment_id = Uuid::new_v4();
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug.clone(),
+            status: EnrollmentStatus::Enrolled {
+                enrollment_id,
+                branch: "control".to_owned(),
+                reason: EnrolledReason::Qualified,
+            },
+        };
+        let enrollment = evolver
+            .evolve_enrollment(
+                true,
+                Some(&exp),
+                Some(&exp),
+                Some(&existing_enrollment),
+                &mut events,
+            )?
+            .unwrap();
+        if let EnrollmentStatus::Disqualified {
```

Comment:
As noted in a comment above, I think the expected behaviour here is that we stay in the same branch rather than becoming disqualified. (As opposed to if the *targetting* had changed, which would disqualify us).